File: clients/java/src/main/java/com/themisdb/client/CircuitBreaker.java
package com.themisdb.client;

import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Circuit Breaker implementation for fault tolerance
 * 
 * Implements the circuit breaker pattern to prevent cascading failures
 * by temporarily blocking requests when a service is experiencing issues.
 */
public class CircuitBreaker {
    
    /**
     * Circuit breaker states
     */
    public enum State {
        CLOSED,      // Normal operation
        OPEN,        // Blocking requests due to failures
        HALF_OPEN    // Testing if service has recovered
    }
    
    private final AtomicReference<State> state = new AtomicReference<>(State.CLOSED);
    private final AtomicInteger failureCount = new AtomicInteger(0);
    private final AtomicInteger successCount = new AtomicInteger(0);
    private final AtomicReference<Instant> nextAttemptTime = new AtomicReference<>(Instant.now());
    
    private final int failureThreshold;
    private final Duration resetTimeout;
    private final int halfOpenMaxRequests;
    
    /**
     * Create a new circuit breaker
     * 
     * @param failureThreshold Number of failures before opening circuit
     * @param resetTimeout How long to wait before attempting reset
     * @param halfOpenMaxRequests Max requests in half-open state
     */
    public CircuitBreaker(int failureThreshold, Duration resetTimeout, int halfOpenMaxRequests) {
        this.failureThreshold = failureThreshold;
        this.resetTimeout = resetTimeout;
        this.halfOpenMaxRequests = halfOpenMaxRequests;
    }
    
    /**
     * Check if a request can be executed
     * 
     * @return true if request should proceed, false if blocked
     */
    public boolean canExecute() {
        State currentState = state.get();
        
        if (currentState == State.CLOSED) {
            return true;
        }
        
        if (currentState == State.OPEN) {
            if (Instant.now().isAfter(nextAttemptTime.get())) {
                transitionToHalfOpen();
                return true;
            }
            return false;
        }
        
        // HALF_OPEN state
        return successCount.get() < halfOpenMaxRequests;
    }
    
    /**
     * Record a successful request
     */
    public void recordSuccess() {
        State currentState = state.get();
        
        if (currentState == State.HALF_OPEN) {
            int count = successCount.incrementAndGet();
            if (count >= halfOpenMaxRequests) {
                state.set(State.CLOSED);
                failureCount.set(0);
            }
        } else if (currentState == State.CLOSED) {
            failureCount.set(0);
        }
    }
    
    /**
     * Record a failed request
     */
    public void recordFailure() {
        int count = failureCount.incrementAndGet();
        if (count >= failureThreshold) {
            state.set(State.OPEN);
            nextAttemptTime.set(Instant.now().plus(resetTimeout));
        }
    }
    
    /**
     * Transition to half-open state
     */
    private void transitionToHalfOpen() {
        state.set(State.HALF_OPEN);
        successCount.set(0);
    }
    
    /**
     * Get the current circuit breaker state
     * 
     * @return Current state
     */
    public State getState() {
        return state.get();
    }
}

File: clients/csharp/ThemisDB.Client/CircuitBreaker.cs
/*
╔═════════════════════════════════════════════════════════════════════╗
║ ThemisDB - Hybrid Database System                                   ║
╠═════════════════════════════════════════════════════════════════════╣
  File:            CircuitBreaker.cs                                  ║
  Version:         0.0.45                                             ║
  Last Modified:   2026-04-15 07:05:23                                ║
  Author:          unknown                                            ║
╠═════════════════════════════════════════════════════════════════════╣
  Quality Metrics:                                                    ║
    • Maturity Level:  🟢 PRODUCTION-READY                             ║
    • Quality Score:   100.0/100                                      ║
    • Total Lines:     172                                            ║
    • Open Issues:     TODOs: 0, Stubs: 0                             ║
╠═════════════════════════════════════════════════════════════════════╣
  Revision History:                                                   ║
    • 2a1fb04231  2026-03-03  Merge branch 'develop' into copilot/audit-src-module-docu... ║
╠═════════════════════════════════════════════════════════════════════╣
  Status: ✅ Production Ready                                          ║
╚═════════════════════════════════════════════════════════════════════╝
 */

using System;

namespace ThemisDB.Client;

/// <summary>
/// Circuit breaker implementation for fault tolerance
/// Prevents cascading failures by temporarily blocking requests when a service is experiencing issues
/// </summary>
public class CircuitBreaker
{
    private readonly int _failureThreshold;
    private readonly TimeSpan _resetTimeout;
    private readonly int _halfOpenMaxRequests;
    private readonly SemaphoreSlim _lock = new(1, 1);

    private CircuitBreakerState _state = CircuitBreakerState.Closed;
    private int _failureCount;
    private int _successCount;
    private DateTime _nextAttemptTime;

    /// <summary>
    /// Circuit breaker states
    /// </summary>
    public enum CircuitBreakerState
    {
        Closed,
        Open,
        HalfOpen
    }

    /// <summary>
    /// Initializes a new circuit breaker
    /// </summary>
    public CircuitBreaker(int failureThreshold, TimeSpan resetTimeout, int halfOpenMaxRequests)
    {
        _failureThreshold = failureThreshold;
        _resetTimeout = resetTimeout;
        _halfOpenMaxRequests = halfOpenMaxRequests;
        _nextAttemptTime = DateTime.MinValue;
    }

    /// <summary>
    /// Current state of the circuit breaker
    /// </summary>
    public CircuitBreakerState State => _state;

    /// <summary>
    /// Check if a request can be executed
    /// </summary>
    public async Task<bool> CanExecuteAsync()
    {
        await _lock.WaitAsync();
        try
        {
            if (_state == CircuitBreakerState.Closed)
            {
                return true;
            }

            if (_state == CircuitBreakerState.Open)
            {
                if (DateTime.UtcNow >= _nextAttemptTime)
                {
                    _state = CircuitBreakerState.HalfOpen;
                    _successCount = 0;
                    return true;
                }
                return false;
            }

            // Half-open state
            return _successCount < _halfOpenMaxRequests;
        }
        finally
        {
            _lock.Release();
        }
    }

    /// <summary>
    /// Record a successful request
    /// </summary>
    public async Task RecordSuccessAsync()
    {
        await _lock.WaitAsync();
        try
        {
            _failureCount = 0;

            if (_state == CircuitBreakerState.HalfOpen)
            {
                _successCount++;
                if (_successCount >= _halfOpenMaxRequests)
                {
                    _state = CircuitBreakerState.Closed;
                }
            }
        }
        finally
        {
            _lock.Release();
        }
    }

    /// <summary>
    /// Record a failed request
    /// </summary>
    public async Task RecordFailureAsync()
    {
        await _lock.WaitAsync();
        try
        {
            _failureCount++;
            _successCount = 0;

            if (_failureCount >= _failureThreshold)
            {
                _state = CircuitBreakerState.Open;
                _nextAttemptTime = DateTime.UtcNow.Add(_resetTimeout);
            }
        }
        finally
        {
            _lock.Release();
        }
    }

    /// <summary>
    /// Execute a function with circuit breaker protection
    /// </summary>
    public async Task<T> ExecuteAsync<T>(Func<Task<T>> action)
    {
        if (!await CanExecuteAsync())
        {
            throw new InvalidOperationException("Circuit breaker is OPEN");
        }

        try
        {
            var result = await action();
            await RecordSuccessAsync();
            return result;
        }
        catch
        {
            await RecordFailureAsync();
            throw;
        }
    }
}
